lib/improver/weights.py

```diff
@@ -30,123 +30,259 @@
+
+        weights = WeightsUtilities.normalise_weights(np.array(weights_list))
+
+        return weights
+
+    @staticmethod
+    def process_coord(cube, coord):
+        """Calculated weights for a given cube and coord.
+
+            Args:
+                cube : iris.cube.Cube
+                       Cube to blend across the coord.
+                coord : iris.corrd.Coord
+                        Coordinate in the cube to be blended.
+            Returns:
+                weights :  (numpy.ndarray) of weights
+                           where sum = 1.0
+                           and missing weights are set to -1.0
+                           size is the same as points in coord if
+                           points set in coord otherwise defaults to
+                           point in cube.coord
+        """
+        if not isinstance(coord, iris.coords.Coord):
+            msg = ('The second argument must be an instance of '
+                   'iris.coords.Coord but is'
+                   ' {0:s}'.format(type(coord)))
+            raise ValueError(msg)
+
+        req_coord_name = str(coord.name())
+        if not cube.coords(req_coord_name):
+            msg = ('The coord for this plugin must be '
+                   'an existing coordinate in the input cube.')
+            raise ValueError(msg)
+
+        cube_coord = cube.coord(req_coord_name)
+
+        if len(coord.points) == 0:
+            # coord points unset will default to all points in cube
+            num_of_weights = len(cube_coord.points)
+        elif len(coord.points) < len(cube_coord.points):
+            msg = ('The cube coordinate has more points '
+                   'than requested coord, '
+                   'len coord points = {0:d} '.format(len(coord.points)) +
+                   'len cube points = {0:d}'.format(len(cube_coord.points)))
+            raise ValueError(msg)
+        else:
+            num_of_weights = len(coord.points)
+
+        # Find missing points
+        missing_weights = np.ones(num_of_weights)
+        if len(cube_coord.points < num_of_weights):
+            # Firstly check that coord is in the right units
+            # Do not try if coord.units not set
+            if (coord.units != cf_units.Unit('1') and
+                    str(coord.units) != 'no_unit'):
+                if coord.units != cube_coord.units:
+                    try:
+                        coord.convert_units(cube_coord.units)
+                    except:
+                        msg = ('Failed to convert coord units '
+                               'requested coord units '
+                               '= {0:s} '.format(str(coord.units)) +
+                               'cube units '
+                               '= {0:s}'.format(str(cube_coord.units)))
+                        raise ValueError(msg)
+            for i, val in enumerate(coord.points):
+                if val not in cube_coord.points:
+                    missing_weights[i] = 0.0
```

Comment:
I've renamed this `forecast_present` which I think is clearer, but am open to other suggestions. I've also rename `missing_sum` to `num_forecast_present' which I think illustrates its purpose better.